### DISCUSSION TITLE
Feature Request

### DISCUSSION THREAD

**Discussion by NexGen-3D-Printing** [HussX/piplay#10] [Ideas]
Hi HussX,

Firstly I just wanted to say, awesome work, this is an excellent little project, I have spent 2 days trying all sorts of things, just to display Birdseye view from Frigate using the direct RTSP feed, and everything else just falls over and stops working on a RPI 4/5, this works perfectly, its simple, its fast and its stable.

Two features to request:

1.  Maybe add a discussion section to you repo.
2.  Maybe as a future addition, you can add an auto rotation for each RTSP feed?

**Comment by HussX** [comment]
That may be easy enough to implement as I get time. I'll definitely look
into the discussion addition. Just maybe add a rotation flag to the config.
I need to create an update script before I make many more other changes
since I relocate files after install. I have a decent logic for it in my
mind if I don't forget by the time I can get to it.

**Comment by NexGen-3D-Printing** [comment]
All good, I just tested today on a 2GB Pi 5, 9 x 720p 15fps camera feeds directly and its perfect, better than using Frigates Birdseye view, I'm really impressed, just wondering if I can get this working on an x86 board using Debian or Ubuntu server, I also have some other Arm boards than run Armbian, I might test on one of those as well.

The Pi 4 seems to struggle a little, could be its wifi though, I need to test over ethernet, but the Pi 5 is actually quite could, first time I've seen it justify the additional cost :)

**Comment by HussX** [comment]
I have a Lepotato in my office I could test on. Most of the code would be
identical. But I'm about 1000 miles from my office as of today.

**Comment by NexGen-3D-Printing** [comment]
Okay, just tested on Ubuntu Server 24 on a Dell Latitude 7235 tablet running a 7th gen Intel M processor, and its working, but the install script failed, had an issue with a hash mismatch or something and stopped, so I installed the following packages myself, and run the install script again:

sudo apt install python3-opencv
sudo apt install python3-pyqt6
sudo apt install gstreamer1.0-plugins-good
sudo apt install gstreamer1.0-plugins-bad
sudo apt install gstreamer1.0-plugins-ugly
sudo apt install python3-yaml

After this, it fired straight up and is working perfectly.

**Comment by NexGen-3D-Printing** [comment]
I can see this software being perfect if you want a security camera display on the wall using an old useless Surface tablet or similar device, I have an old Getac tablet with a 4th gen I5 in it, I might see how many h264 feeds it can handle before it falls over, this software should be really handy for upcycling on tablet PC's, no point spending money on an underpower SBC and screen if you can pick up a complete tablet for less money and run Ubuntu Server on it.

Below is the Pi 5 running on a Samsung Screen:

![IMG_20241104_165551_621](https://github.com/user-attachments/assets/9cd743e2-ce2a-4981-be9d-b260c0561bbc)

**Comment by SvenVD** [comment]
> I can see this software being perfect if you want a security camera display on the wall using an old useless Surface tablet or similar device, I have an old Getac tablet with a 4th gen I5 in it, I might see how many h264 feeds it can handle before it falls over, this software should be really handy for upcycling on tablet PC's, no point spending money on an underpower SBC and screen if you can pick up a complete tablet for less money and run Ubuntu Server on it.
> 
> Below is the Pi 5 running on a Samsung Screen:
> 
> ![IMG_20241104_165551_621](https://private-user-images.githubusercontent.com/20825418/382696837-9cd743e2-ce2a-4981-be9d-b260c0561bbc.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NjU5MjQsIm5iZiI6MTczMDc2NTYyNCwicGF0aCI6Ii8yMDgyNTQxOC8zODI2OTY4MzctOWNkNzQzZTItY2UyYS00OTgxLWJlOWQtYjI2MGMwNTYxYmJjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA1VDAwMTM0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0MTNkZDczNjViNjI5OGQ5NGUxZTZlOWYzNGVlYzQzMmNhNzBhNDNkNTI2NTllYTg1N2EyZThkZGI0NDBiMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VD1EDVcLYq6u5JD5H-L1-JqoqxiaoNC3vhkBV1ZlRmo)

May I ask, what is the latency of the streams and what resolutions and fps are the streams on the raspberry pi 5?

**Reply by NexGen-3D-Printing** [reply]
Each of my feeds are 720p H264 2mbit 15fps, latency is really good, almost instant, slight delay due to normal network related things, would be better if running over ethernet though.

The screens resolution is 1080p

I might drop the resolution of each feed and see if I can run a higher frame rate, as anything above 15fps is just to much for the Pi 5, its probably not running at the full 15fps as it is, but there is no way to tell, it looks about right, maybe somewhere between 10 and 15.

**Reply by HussX** [reply]
If you get a wild hair, try out the piplayAlt.py and main.qml I uploaded.  I haven't pushed it into the startup script yet but I'm thinking of doing a lateral move.  Some additional apt-get installs are listed at the top of piplayAlt.py and all you need to do is paste that into the regular piplay.py and restart the service, so long as main.qml is in the same directory.  I haven't removed the cprofile yet since I'm still tweaking, but it cut the CPU use on a Pi 3 over half.  If it works for you on other platforms as well, I'm going to start testing a new round of what's really needed from apt.

**Comment by HussX** [comment]
> > I can see this software being perfect if you want a security camera display on the wall using an old useless Surface tablet or similar device, I have an old Getac tablet with a 4th gen I5 in it, I might see how many h264 feeds it can handle before it falls over, this software should be really handy for upcycling on tablet PC's, no point spending money on an underpower SBC and screen if you can pick up a complete tablet for less money and run Ubuntu Server on it.
> > Below is the Pi 5 running on a Samsung Screen:
> > ![IMG_20241104_165551_621](https://private-user-images.githubusercontent.com/20825418/382696837-9cd743e2-ce2a-4981-be9d-b260c0561bbc.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3NjU5MjQsIm5iZiI6MTczMDc2NTYyNCwicGF0aCI6Ii8yMDgyNTQxOC8zODI2OTY4MzctOWNkNzQzZTItY2UyYS00OTgxLWJlOWQtYjI2MGMwNTYxYmJjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA1VDAwMTM0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0MTNkZDczNjViNjI5OGQ5NGUxZTZlOWYzNGVlYzQzMmNhNzBhNDNkNTI2NTllYTg1N2EyZThkZGI0NDBiMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VD1EDVcLYq6u5JD5H-L1-JqoqxiaoNC3vhkBV1ZlRmo)
> 
> May I ask, what is the latency of the streams and what resolutions and fps are the streams on the raspberry pi 5?

@NexGen-3D-Printing that's some bright green grass! My use case was always 2-4 cameras so it works out great for me either way and I always have enough PIs on hand to repurpose if needed. I have countless PI projects at work 😂 and a 3x2 at home. 
@SvenVD I don't know if there's a perfect way to answer that question. As with any other streaming program it's a good idea to keep downsampling in mind or else you're making the board work extra for no reason. Ie if you're running 2x2 on a 1920x1080 screen, you don't really want the substream you're sourcing to be larger than the 960x540 it's using on the screen. This isn't technically playing the videos, but drawing them into frame "photos" sized to the space they're actually using and dropping frames when they're behind the buffer. So playing a 4k camera in that case would still result in a 960x540 view, but with a lot of unnecessary bandwidth. It's best to try and closely match a substream to the space it'll be using so your PI doesn't have to fight to keep up. 
I never directly measured the latency but it will mostly depend on whether you're direct/remote and wired/Wi-Fi. Since it drops outdated frames, it really shouldn't ever be more than a couple seconds behind but YMMV.

**Comment by HussX** [comment]
@NexGen-3D-Printing @SvenVD FYI, moved to discussion. In case it breaks email responses.

**Comment by SvenVD** [comment]
I asked because we are running into an issue trying to use mpv in X11 to display RTSP streams on a raspberry pi 5 https://github.com/SvenVD/rpisurv/issues/194 so I was wondering what is the difference that mpv can not handle but opencv seemingly can. Maybe we are missing some mpv options

**Comment by HussX** [comment]
Oof, apologies, I thought your name looked familiar. You already knew what I typed out and then some. 😂 I tried rpisurv, but I wanted something smooth for native "Raspbian". He can probably answer that better as I didn't try testing upper limits with resolution and I have nothing at my location this week to really try with.
